### ISSUE TITLE
Setup client-side declared grid on server

### ISSUE THREAD

**Issue by alexberazouski** [vaadin/flow#2817]
Trying to setup vaadin-grid:

On client side:
`<vaadin-grid id="users-grid" theme="borderless"></vaadin-grid>`

On server: 
```
@Id("users-grid")
private Grid<User> grid;
```
and also I define some columns and dataProvider

After I open the page with grid on client I receive following errors:
```(TypeError) : Cannot read property '$connector' of undefined```
```(TypeError) : Cannot set property 'size' of undefined```

Note that: 
- On client I see all server defined columns
- Header cells contain proper column names
- Other cells contain “” or “ “ or template in case of TemplateRendered column, but anyway it’s not displayed correctly.

As a workaround - it works if I create a grid instance on server and put it into slot

**Comment by pleku** [comment]
Setting a timeboxed investigation of 1 for now.

**Comment by gilberto-torrezan** [comment]
This is probably caused by an underlying problem we have when using `@Id` mappings:
The code that is not working on Grid is:
```
getElement().getNode()
                .runWhenAttached(ui -> ui.getPage().executeJavaScript(
                        "window.gridConnector.initLazy($0)", getElement()));
```
This is not working because the `getElement()` doesn't have a `nodeId` at the constructor of the Grid when using `@Id`, so the client engine can't find the exact node to send to the javascript call - leading to the `Cannot read property ... of undefined` errors

**Comment by denis-anisimov** [comment]
Generally Gilberto is right, but the accurate reason is here:
- The injected `Element` server-side instance has no real client-side counterpart when it's instantiated (actually plain "non-injected" element also has no this counterpart but since all JS invocations happens _AFTER_ attach and bound actions (they are deferred) it will have it once JS is really executed).
- It will get the client side counterpart only after round-trip. After this roundtrip the fake serves-side `StateNode` becomes a real `StateNode` which has the `StateNode` client side counterpart with `domeNode` (DOM element) assigned to this client-side `StateNode` instance.
- Only when the client-side `StateNode` instance gets its DOM element it's possible to call JS function from the server side (now the client side element is addressed via a `StateNode`).
- But `Grid` calls JS code in its CTOR where this is not yet true and it fails to initialize its connector.

Now what can be done:
- This is quite generic issue: each time when we have a component which is injected via `@Id` this situation may occur.
- It can be decided either on the server-side or the client-side.
- On the server side it might be that we should not fire an attach event at all until the element has no its client-side counterpart (so the "attach" may be considered from the client-side point of view). See the meta-code below.
- On the client side we can introduce additional logic which looks for the DOM element not only via `StateNode::getDomNode` but also via some features (like `com.vaadin.client.ExistingElementMap`).

Here is the code that can be used on the server side as a generic solution to avoid the issue:

`StateNode` : `private void handleOnAttach() {`
```
...
//replace fireAttachListeners(initialAttach); to
if (getParent() != null
                && getParent().hasFeature(AttachTemplateChildFeature.class)) {
            AttachTemplateChildFeature feature = getParent()
                    .getFeature(AttachTemplateChildFeature.class);
            Element element = feature.getParent(this);
            if (element == null) {
                fireAttachListeners(initialAttach);
            }
        } else {
            fireAttachListeners(initialAttach);
        }
```

This is bad because `StateNode` shouldn't know anything about any specific feature.
But that's just an idea. The code could be modified so that it does the same but using a correct way.
It can't be a good solution anyway since we need to "postpone" attach event depending on the specific feature behavior.

I would say that it's better to do it on the client side. It will require additional logic there as well which will also depends on the specific feature. But it won't be a generic code and it doesn't look so bad on the client side.
The only thing that would be nice to care here: JS code execution should be aware of server-side roundtrip. It might be that it's safe to execute the code right away without waiting the answer from the server side (when the node becomes a part of the `VirtualChildrenList` feature). Or it might be that it should be somehow postponed to wait until it happens.

**Comment by denis-anisimov** [comment]
Just a note:

with the server-side code above for `StateNode` `Grid` works fine being injected by `@Id`.
I've checked that.

**Comment by denis-anisimov** [comment]
My personal proposal is:

modify `ExecuteJavaScriptProcessor` so that it has internal queue of invocations (queue of `JsonArray`). Each invocation is represented by `JsonArray`.

No when `execute` is called (*NB* : make sure it's called even if there is JS executions requested. `MessageHandler::processMessage` should check whether the internal queue is empty) the `ExecuteJavaScriptProcessor` instance should go first through its internal queue in the same way as with the new executions.

For every execution: check whether there a Dom Node  available for every `StateNode` parameter.
If it's available then execute it immediately. If not then check whether `ExistingElementMap`  has an `Element` by the  node id. If it's there then don't execute the execution. Add it into internal queue.

The internal queue should be emptied each time. And it should be filled again with the executions which are required to be postponed. It will allow to remove those executions that has wrong `StatNode`'s id (and shouldn't be executed at all).

**Comment by Legioth** [comment]
Please keep in mind that there might also be cases when it's desirable that the RPC is executed with a `null` parameter in cases when the some element or node is no longer attached.

I think the proper solution here would be to remove the asynchronous step with `@Id` mappings. This would in theory not be 100% correct in cases where the server would assume an element exists even though it couldn't be found on the client. On the other hand, it would help reduce lots of complexity since in practice, nobody is doing anything to handle the case when the expected element couldn't be located.

**Comment by denis-anisimov** [comment]
Async step with `@Id` mapping cannot be removed completely: we still need to notify server side about failure in case there is no requested `id` and we need also track somehow "id" children which are not children at all.
The only thing that is not required is: postpone attaching DOM Element to the `StateNode`.
That can be done right away, yes. And it will fix the issue.

Cleaning the logic which handles DOM element asynchronously may be extracted as a different issue.

**Comment by denis-anisimov** [comment]
I've been investigating the fix on the client side.
Important to note that JS execution in Grid CTOR happens _BEFORE_ even the element is requested to be attached to the server side (via the `attachExistingElementById` method call).
It means that the element can't be eligible for any JS call regardless of our decision how to fix this.

To change the JS call order we should change the current call in the `TemplateInitializer` class
from 
```
templateNode.runWhenAttached(ui -> {
                templateNode.getFeature(AttachTemplateChildFeature.class)
                .register(getElement(), proposedNode);
                ui.getPage().executeJavaScript(
                        "this.attachExistingElementById($0, $1, $2, $3);",
                        getElement(), tagName, proposedNode.getId(), id);
            });
```
to 
```
proposedNode.runWhenAttached(ui -> ui.getPage().executeJavaScript(
                    "this.attachExistingElementById($0, $1, $2, $3);",
                    getElement(), tagName, proposedNode.getId(), id));

            templateNode.runWhenAttached(ui -> templateNode
                    .getFeature(AttachTemplateChildFeature.class)
                    .register(getElement(), proposedNode));
```

(and everywhere else in the class).

Another note is about my comment "JS code execution should be aware of server-side roundtrip. It might be that it's safe to execute the code right away without waiting the answer from the server side".
It's not safe to execute the JS code without waiting the answer from the server.
JS code may want to use JS additions to the DOM element (like `.$server` , etc) which appears only after the node becomes bound on the client side. Binding happens only after the server round-trip.
It can be seen on the `Grid` example which tries to call server-side immediately in `lazyInit` JS call:
`grid.$server.setRequestedRange` . There is no yet `$server`  property available and this fails because element is not yet bound.

So the options are:
- Bind the `StateNode` right away without server round-trip as Leif suggested. It will require a lot of changes because we are using two `Features` to track "existing element attachment" functionality. We won't need this anymore but the logic will need to be updated keeping in mind unhappy path (there is no such element on the client side and it needs to be removed on the server side. Which means that we still needs server round-trip but it's cleaning up roundtrip).
- Do it on the client side with internal JS execution queue.
- Do it on the server side using meta code which I've written here in the very beginning.

Now I would say that I prefer the last choice using specific approach which eliminate the need to know `StateNode` anything about its features.
It requires less changes.

**Comment by denis-anisimov** [comment]
The server side approach makes the code less fragile: there is no need to be sure about the order of JS calls.
Existing element won't be attached until it's completely initialized which can't happen before `attachExistingElementById`.

**Comment by denis-anisimov** [comment]
Finally I suggest:
- do it in the way which requires less number of changes (modified suggestion with the fix on the client side: add post flush listener to postpone JS execution until the node is not in the `ExistingElementMap`.
- wait until upcoming client side communication refactoring is done which may help to solve this issue more properly.

**Comment by Legioth** [comment]
I agree that it might make sense to postpone the proper fix until after refactoring the client side communication.

On the other hand, I'm worried about leaving any JS executions in a queue if a referred element doesn't (yet) exist. This might lead to unintended results in cases when the element is actually optional and the JS execution is prepared to deal with the value being `undefined`. The server-side caller would thus need some way of signalling how they want it to work.

**Comment by denis-anisimov** [comment]
I understand you concern.
But please note that I'm not going to push an execution into the queue based on the fact that `StateNode` has no `DomNode` (like detached element).
The execution will be postponed _ONLY_ if there is a `StateNode` which is in the `ExistingElementMap`. This map tracks _ONLY_ elements which should get confirmation from the server for final binding. 
Once this confirmation is received they are removed from there and get their `DomNode` (+ properly bound).
Otherwise they are removed from the map.

As a result only 3 cases are possible:
- the node has a DomNode. It's bound and is eligible as a JS argument.
- the node has no DomNode but it's in the `ExistingElementMap`. The execution will be postponed. Later on it will be removed from the map and one of other two cases will be only valid during execution.
- the node has no DomNode and it's not in the `ExistingElementMap`. Process the execution as usual.

**Comment by Legioth** [comment]
Good point. This still leaves a small opportunity for surprises, but I think it's acceptable in this case.

**Comment by Legioth** [comment]
Could test #2817 again once this has been fixed.
